lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,278 @@
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def ip_exists(ip, subnet, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse get("/subnet/#{subnet}/#{ip}?group=#{URI.escape(group.to_s)}")
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
+      response['code'] == 200
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain IP address for subnet_id %s from External IPAM."), subnet)
+    end
+
+    # Deletes an IP address from a given subnet in External IPAM.
+    #
+    # Inputs: 1. ip:               IP address to be freed up
+    #         2. subnet:           The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24",
+    #                              IPv6 - "2001:db8:abcd:12::/124")
+    #         3. group(optional):  The name of the External IPAM group containing the subnet.
+    #
+    # Returns: true if IP is deleted successfully from External IPAM, otherwise false.
+    #
+    # Proxy responses:
+    #   Response if success:
+    #     {"code": 200}
+    #   Response if subnet error:
+    #     {"code": 404, "error": "The specified subnet does not exist in External IPAM."}
+    #   Response if IP already deleted:
+    #     {"code": 404, "error": "No addresses found"}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def delete_ip_from_subnet(ip, subnet, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse delete("/subnet/#{subnet}/#{ip}?group=#{URI.escape(group.to_s)}")
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
+      response['code'] == 200
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to delete IP %s from the subnet %s in External IPAM."), ip, subnet)
+    end
+
+    # Gets a subnet from a specific External IPAM group.
+    #
+    # Inputs: 1. subnet: The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24", IPv6 - "2001:db8:abcd:12::/124")
+    #         2. group: The name of the External IPAM group containing the subnet.
+    #
+    # Returns: A subnet with a "data" key, or a hash with a "message" key containing error.
+    #
+    # Proxy responses:
+    #   Response if exists:
+    #     {"code": 200, "data": {"subnet":"172.55.55.0", "mask":"24", "description":"My subnet"}
+    #   Response if not exists:
+    #     {"code": 404, "error": "No subnet 172.55.66.0/29 found in section '<:group>'"}
+    def get_subnet_from_group(subnet, group)
+      raise "subnet must be provided" if subnet.blank?
+      raise "group must be provided" if group.blank?
+      parse get("/group/#{URI.escape(group.to_s)}/subnet/#{subnet}")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to check if subnet exists in the External IPAM group."))
```

Comment:
raise ProxyException.new(url, e, N_("Unable to check if subnet %{subnet} exists in the External IPAM group %{group}."), subnet: subnet, group: group)

Reply:
Also this method could also get some error handling code, same as above.